src/coreclr/jit/liveness.cpp

```diff
@@ -1475,6 +1475,13 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
                     // or have a behavioral semantic that is undesirable to remove
                     break;
                 }
+#if defined(TARGET_ARM64)
+                else if (HWIntrinsicInfo::HasImmediateOperandRange(intrinsicId))
```

Comment:
I think this bit is overly conservative. We don't need to preserve any intrinsic with an immediate range, rather only ones that have something we can't determine is in bounds.

Consider for example if it was `ShiftLeftLogicalSaturate(vr0, 1)`.

Reply:
It is conservative, but I think that's ok for now. Given this fixes a correctness issue now, we can figure out how to optimize it later.

Reply:
Normally I'd agree, but this is going to mark a very large number of core APIs as side effecting even when they aren't and will likely have impact to important workloads.

Today, the general logic of "do I need a bounds check" goes through `lookupImmBounds` followed by `addRangeCheckIfNeeded`, so I'd expect us to be able to use this to mark the `GT_HWINTRINSIC` as side effecting and then key this liveness check based on that.

I'd expect if we do that, it also solves the issue of ensuring all nodes which insert bounds checks are marked that way, without needing to use this separate `HasImmediateOperandRange` check

Reply:
For reference, here's where we check in rationalization for which operands are immediate and whether they are in range or not: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/rationalize.cpp#L457-L499

That then helps us decide whether or not we need to convert it back to a user call or if we can handle it "as is".

We have similar overall logic in the importer (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsic.cpp and https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsicarm64.cpp, look for `addRangeCheckIfNeeded`), and I think we can mostly hook into the importation handling here to ensure this all works nicely and tracks the side effect.